plan/eliminate_projection.go

```diff
@@ -14,50 +14,215 @@
 package plan
 
 import (
+	"github.com/pingcap/tidb/context"
 	"github.com/pingcap/tidb/expression"
 )
 
-// EliminateProjection eliminates projection operator to avoid the cost of memory copy in the iterator of projection.
-func EliminateProjection(p PhysicalPlan) PhysicalPlan {
-	switch plan := p.(type) {
-	case *Projection:
-		if !projectionCanBeEliminated(plan) {
-			break
-		}
-		child := plan.children[0].(PhysicalPlan)
-		child.SetSchema(plan.Schema())
-		RemovePlan(p)
-		p = EliminateProjection(child)
-	}
-	children := make([]Plan, 0, len(p.Children()))
-	for _, child := range p.Children() {
-		children = append(children, EliminateProjection(child.(PhysicalPlan)))
+// canLogicalProjectionBeEliminated checks if a PROJECTION in a logical plan can be eliminated.
+// Projection p can be eliminated when all the following conditions are satisfied:
+// 1. it's not the root operator
+// 2. canProjectionBeEliminated(p) returns true
+func canLogicalProjectionBeEliminated(p *Projection) bool {
+	if len(p.Parents()) == 0 {
+		return false
 	}
-	p.SetChildren(children...)
-	return p
+	return canProjectionBeEliminated(p)
 }
 
-// projectionCanBeEliminated checks if a PROJECTION operator can be eliminated.
-// PROJECTION operator can be eliminated when meet the following conditions at the same time:
+// canProjectionBeEliminated checks if a PROJECTION in a physical plan can be eliminated.
+// Projection p can be eliminated when all the following conditions are satisfied:
 // 1. fields of PROJECTION are all columns
 // 2. fields of PROJECTION are just the same as the schema of the child operator (including order, amount, etc.).
 // expressions like following cases can not be eliminated:
 // "SELECT b, a FROM t",
 // or "SELECT c AS a, c AS b FROM t WHERE d = 1",
 // or "SELECT t1.a, t2.b, t1.b, t2.a FROM t1, t2 WHERE t1.a < 0 AND t2.b > 0".
-func projectionCanBeEliminated(p *Projection) bool {
-	child := p.children[0].(PhysicalPlan)
+func canProjectionBeEliminated(p *Projection) bool {
+	child := p.Children()[0]
 	if p.Schema().Len() != child.Schema().Len() {
 		return false
 	}
+	childCols := child.Schema().Columns
 	for i, expr := range p.Exprs {
-		col, ok := expr.(*expression.Column)
-		if !ok {
-			return false
-		}
-		if col.FromID != child.Schema().Columns[i].FromID || col.Position != child.Schema().Columns[i].Position {
+		projCol, ok := expr.(*expression.Column)
+		if !ok || projCol.Identify() != childCols[i].Identify() {
 			return false
 		}
 	}
 	return true
 }
+
+func resolveColumnAndReplace(origin *expression.Column, replace map[string]*expression.Column) {
+	dst := replace[(*origin).Identify()]
+	if dst != nil {
+		*origin = *dst
+	}
+}
+
+func resolveExprAndReplace(origin expression.Expression, replace map[string]*expression.Column) {
```

Comment:
If use `expression.ColumnSubstitute(expr Expression, schema *Schema, newExprs []Expression)`:
1. We have to convert `replace`, whose type is `map[string]*expression.Column`, to a `(schema *Schema`, `newExprs [] Expression)` pair every time before function `expression.ColumnSubstitute` called. It's inconvenient
 2. `expression.ColumnSubstitute` uses a `for` loop to find whether a column exists in `schema *Schema`. It's inefficient

I prefer not to use the function `expression.ColumnSubstitute`